File: p398/p398.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <bitset>
#include <vector>
using namespace std;

int findSecondShortest(const bitset<10000000>& rope) {
	vector<int> lengths;
	int shortest = rope.size();
	int s = rope.size();
	int prevCut = 0;
	for (int i = 0; i <= s; ++i) {
		if (i == s || rope[i]) {
			int length = i - prevCut;
			lengths.push_back(length);
			prevCut = i;
			if (length < shortest)
				shortest = length;
			//cout << shortest << endl;
		}
	}

	int secondShortest = rope.size();
	vector<int>::const_iterator end = lengths.end();
	for (vector<int>::const_iterator it = lengths.begin(); it != end; ++it) {
		if (*it >= shortest && *it <= secondShortest) {
			secondShortest = *it;
		}
	}
//	cout << secondShortest << endl;
	return secondShortest;
}

int main(int argc, char* argv[]) {
	int ropeL = 10000000;
	int points = 100;

	int iterations = 100;
	if (argc > 1)
		iterations = atoi(argv[1]);

//	vector<bool> rope(ropeL);
	bitset<10000000> rope;
	int secondShortestLength = 0;

	srand(time(0));

	for (int i = 0; i < iterations; ++i) {
		for (int j = 0; j < points - 2; ++j) {
			int cutPoint = rand() % (ropeL-1) + 1;
			cout << cutPoint << endl;
			if (rope[cutPoint])
				--j;
			else
				rope[cutPoint] = 1;
		}
		secondShortestLength += findSecondShortest(rope);
		//cout << secondShortestLength << endl;
		rope.reset();
//		rope.clear();
//		rope.resize(ropeL);
	}

	cout << 1.0 * secondShortestLength / iterations << endl;
	return 0;
}

File: p398/p398.java
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;

public class p398 {

	private static final int ROPE_LENGTH = 10000000;
	private static final int CUT_POINTS = 100;
	private static final int TOTAL_ITERATIONS = 50000;

	private static final AtomicLong secondShortestLength = new AtomicLong();
	private static final AtomicLong completedIterations = new AtomicLong();
	private static final List<Thread> threads = new ArrayList<Thread>();

	
	public static void main(String[] args) {
		setupThreads();
		for (Thread thread : threads) {
			try {
				thread.join();
			} catch (InterruptedException iex) { }
			
		}
		
		System.out.println(String.valueOf(1.0 * secondShortestLength.get() / completedIterations.get()));
	}
	
	private static final void setupThreads() {
		int numberOfThreads = Runtime.getRuntime().availableProcessors();
		System.out.println("Running with " + numberOfThreads + " threads...");
		for (int i = 0; i < numberOfThreads; ++i) {
			Thread thread = new Thread(new Runnable() {
				@Override
				public void run() {
					final BitSet rope = new BitSet(ROPE_LENGTH);
					final ArrayList<Integer> cutLengths = new ArrayList<Integer>();
					while (completedIterations.incrementAndGet() < TOTAL_ITERATIONS) {
						for (int j = 0; j < CUT_POINTS; ++j) {
							final int cutPoint = (int)(Math.random() * (ROPE_LENGTH-1)) + 1;
							//System.out.println(cutPoint + "");
							if (rope.get(cutPoint))
								--j;
							else
								rope.set(cutPoint);
						}
						secondShortestLength.getAndAdd(findSecondShortest(rope, cutLengths));
						//System.out.println(secondShortestLength + "");
						rope.clear();
						cutLengths.clear();
					}
				}
			});
			threads.add(thread);
		}
		for (Thread thread : threads) {
			thread.start();
		}
	}
	
	private static final int findSecondShortest(final BitSet rope, final ArrayList<Integer> cutLengths) {
		int shortest = rope.size();
		int prevCut = 0;
		for (int i = 0; i <= ROPE_LENGTH; ++i) {
			if (i == ROPE_LENGTH || rope.get(i)) {
				int length = i - prevCut;
				cutLengths.add(length);
				prevCut = i;
				if (length < shortest)
					shortest = length;
					
				//System.out.println(shortest + "");
			}
		}

		final int lengthsLen = cutLengths.size();
		int secondShortest = ROPE_LENGTH;
		for (int i = 0; i < lengthsLen; ++i) {
			int length = cutLengths.get(i);
			if (length >= shortest && length <= secondShortest) {
				secondShortest = length;
			}
		}
		return secondShortest;
	}
	
}
